src/components/molecules/TablesUserList/TablesUserList.tsx

```diff
@@ -44,169 +42,189 @@ const firestoreUpdate = (doc: string, update: any) => {
-
-    const update = {
-      data: {
-        ...existingData,
-        [venueName]: {
-          table,
-          videoRoom,
+  // @debt refactor this into api/* layer or similar?
+  const takeSeat = useCallback(
+    (table: string) => {
+      if (!user) return;
+
+      const doc = `users/${user.uid}`;
+      const existingData = recentVenueUsers.find((u) => u.id === user.uid)
+        ?.data;
+
+      const update = {
+        data: {
+          ...existingData,
+          [venueName]: {
+            table,
+            videoRoom,
+          },
         },
-      },
-    };
-    firestoreUpdate(doc, update);
-  };
+      };
+
+      firestoreUpdate(doc, update);
+    },
+    [recentVenueUsers, user, venueName, videoRoom]
+  );
+
+  const tableLocked = useCallback(
+    (table: string) => {
+      const areUsersAtTable = recentVenueUsers.some(
+        (user: User) => getUserExperience(venueName)(user)?.table === table
+      );
 
-  const usersAtOtherTables = [];
-  for (const table of tables) {
-    if (table.reference === seatedAtTable) {
-      continue;
-    }
-    usersAtOtherTables.push(...usersAtTables[table.reference]);
-  }
+      // Empty tables are never locked
+      if (!areUsersAtTable) return false;
+
+      // Locked state is in the experience record
+      return isTruthy(experience?.tables?.[table]?.locked);
+    },
+    [experience?.tables, recentVenueUsers, venueName]
+  );
+
+  const onAcceptJoinMessage = useCallback(
+    (table: string) => {
+      window.scrollTo(0, 0);
+      hideJoinMessage();
+      takeSeat(table);
+      setSeatedAtTable(table);
+    },
+    [hideJoinMessage, setSeatedAtTable, takeSeat]
+  );
+
+  const acceptJoiningTable = useCallback(
+    () => onAcceptJoinMessage(joiningTable),
+    [joiningTable, onAcceptJoinMessage]
+  );
+
+  const onJoinClicked = useCallback(
+    (table: string, locked: boolean, videoRoom: string) => {
+      if (locked) {
+        showLockedMessage();
+      } else {
+        setJoiningTable(table);
+        setVideoRoom(videoRoom);
+        joinMessage ? showJoinMessage() : onAcceptJoinMessage(table);
+      }
+    },
+    [joinMessage, onAcceptJoinMessage, showJoinMessage, showLockedMessage]
+  );
+
+  const renderedTables = useMemo(() => {
+    if (isSeatedAtTable) return;
+
+    return tables.map((table: Table, index: number) => (
+      <TableComponent
+        key={table.reference}
+        experienceName={venueName}
+        users={recentVenueUsers}
+        table={table}
+        tableLocked={tableLocked}
+        onJoinClicked={onJoinClicked}
+        // @debt should this be using the table.reference (rather than index) for nameOfVideoRoom?
+        nameOfVideoRoom={`${venueName}-table${index + 1}`}
+      />
+    ));
+  }, [
+    TableComponent,
+    isSeatedAtTable,
+    onJoinClicked,
+    recentVenueUsers,
+    tableLocked,
+    tables,
+    venueName,
+  ]);
+
+  if (!isRecentVenueUsersLoaded) return <>Loading...</>;
```

Comment:
We should probably use `<LoadingPage/>` or `<Loading/>` here.

Reply:
Agreed. Would have to look closer at the specifics to decide which makes most sense, or which can provide the least disruptive feel for the user.

Reply:
It's rendered within a bunch of other JSX, so definitely `<Loading/>`

Reply:
fixed in 247bc9ac730dcc377ff5e95e5343deaa0f9a1d4e